File: lection04/app_java/src/main/java/ru/hse/lection04/presentationlayer/MainActivity.java
package ru.hse.lection04.presentationlayer;

import android.content.Intent;
import android.os.Bundle;
import android.widget.CompoundButton;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SwitchCompat;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import ru.hse.lection04.R;
import ru.hse.lection04.businesslayer.LogProvider;
import ru.hse.lection04.businesslayer.ServiceLocator;
import ru.hse.lection04.businesslayer.connectivity.AbstractConnectivityProvider;
import ru.hse.lection04.objects.LogEntry;
import ru.hse.lection04.presentationlayer.adapter.LogAdapter;

public class MainActivity extends AppCompatActivity {
    public static final String TAG = "MainActivity";


    protected AbstractConnectivityProvider mConnectivityProvider = ServiceLocator.getConnectivityProvider();
    protected LogProvider mLogProvider = ServiceLocator.getLogProvider();

    protected final LogListener mLogListener = new LogListener();
    protected final LogAdapter mLogAdapter = new LogAdapter();

    protected SwitchCompat mTrackConnectivity;
    protected RecyclerView mRecycler;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        mTrackConnectivity = findViewById(R.id.connectivity_track);
        mTrackConnectivity.setChecked(mConnectivityProvider.isTrackEnabled());
        mTrackConnectivity.setOnCheckedChangeListener(new TrackerChangeListener());

        mRecycler = findViewById(R.id.recycler);
        mRecycler.setLayoutManager(new LinearLayoutManager(this));
        mRecycler.setAdapter(mLogAdapter);

    }

    @Override
    protected void onStart() {
        super.onStart();

        mLogProvider.register(mLogListener);
    }

    @Override
    protected void onResume() {
        super.onResume();

        checkAndUpdateState();
    }

    @Override
    protected void onStop() {
        mLogProvider.unregister(mLogListener);

        super.onStop();
    }


    protected void checkAndUpdateState() {
        if (mTrackConnectivity.isChecked() != mConnectivityProvider.isTrackEnabled()) {
            mTrackConnectivity.setChecked(mConnectivityProvider.isTrackEnabled());
        } else {
            updateServiceState(mConnectivityProvider.isTrackEnabled());
        }
    }

    protected void updateServiceState(boolean isChecked) {
        final Intent intent = ConnectivityService.newInstance(getApplicationContext());

        if (isChecked) {
            ContextCompat.startForegroundService(getApplicationContext(), intent);
        } else {
            stopService(intent);
        }
    }

    protected void updateData(List<LogEntry> entries) {
        mLogAdapter.submitList(entries);
    }


    protected class LogListener implements LogProvider.IListener {
        @Override
        public void logUpdated(LogProvider provider, List<LogEntry> entries) {
            updateData(entries);
        }
    }

    protected class TrackerChangeListener implements CompoundButton.OnCheckedChangeListener {
        @Override
        public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
            final boolean needServiceUpdate = mConnectivityProvider.setTrackEnabled(isChecked);

            if (needServiceUpdate) {
                updateServiceState(isChecked);
            }
        }
    }
}

File: lection03/app/src/main/java/com/otopba/hello_world/MainActivity.kt
package com.otopba.hello_world

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton

class MainActivity : AppCompatActivity() {

    val movieList = generateMovieList().toMutableList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_activity)
        val rv: RecyclerView = findViewById(R.id.rv)
        val adapter = MovieAdapter(movieList)
        rv.adapter = adapter
        rv.layoutManager = LinearLayoutManager(this)
        val fab: FloatingActionButton = findViewById(R.id.fab)
        fab.setOnClickListener {
            movieList.add(Movie("fghfgh", "fhgfgh", 0))
            adapter.notifyDataSetChanged()
        }
    }

    private fun generateMovieList(): List<Movie> {
        return listOf(
            Movie(
                "Побег из Шоушенка",
                "Оказавшись в тюрьме под названием Шоушенк, он сталкивается с жестокостью и беззаконием, царящими по обе стороны решетки. Каждый, кто попадает в эти стены, становится их рабом до конца жизни",
                R.drawable.movie_1
            ), Movie(
                "Матрица",
                "Жизнь Томаса Андерсона разделена на две части: днём он — самый обычный офисный работник, получающий нагоняи от начальства, а ночью превращается в хакера по имени Нео, и нет места в сети, куда он не смог бы дотянуться",
                R.drawable.movie_2
            ), Movie(
                "Как приручить дракона",
                "Вы узнаете историю подростка Иккинга, которому не слишком близки традиции его героического племени, много лет ведущего войну с драконами",
                R.drawable.movie_3
            ), Movie(
                "12 стульев",
                "Во время революции и последовавшего за ней краткого периода военного коммунизма многие прятали свои ценности как можно надежнее",
                R.drawable.movie_4
            ), Movie(
                "Зеленая книга",
                "Утонченный светский лев, богатый и талантливый музыкант нанимает в качестве водителя и телохранителя человека, который менее всего подходит для этой работы",
                R.drawable.movie_5
            ), Movie(
                "Пираты Карибского моря: Проклятие Черной жемчужины",
                "Жизнь харизматичного авантюриста, капитана Джека Воробья, полная увлекательных приключений, резко меняется, когда его заклятый враг — капитан Барбосса — похищает корабль Джека, Черную Жемчужину, а затем нападает на Порт Ройал и крадет прекрасную дочь губернатора, Элизабет Свонн.",
                R.drawable.movie_6
            ), Movie(
                "Гарри Поттер и философский камень",
                "Жизнь десятилетнего Гарри Поттера нельзя назвать сладкой: его родители умерли, едва ему исполнился год, а от дяди и тётки, взявших сироту на воспитание, достаются лишь тычки да подзатыльники",
                R.drawable.movie_7
            )
        )
    }

}